### PR TITLE
Fixes Snackbar and Toast on iOS release builds

### LINKED ISSUE #1752: [BUG] Toast.Show crashing on iOS Release builds
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

When attempting to show a Toast notification, an exception is thrown and no toast is displayed.
```System.InvalidCastException: Arg_InvalidCastException
   at System.Collections.Frozen.SmallFrozenSet`1.GSW[[UIKit.UIView, Microsoft.iOS, Version=17.2.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]].Store(FrozenSet`1 )
   at System.Collections.Frozen.FrozenSetInternalBase`2[[UIKit.UIView, Microsoft.iOS, Version=17.2.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065],[System.Collections.Frozen.SmallFrozenSet`1.GSW[[UIKit.UIView, Microsoft.iOS, Version=17.2.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]], System.Collections.Immutable, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(IEqualityComparer`1 )
   at System.Collections.Frozen.SmallFrozenSet`1[[UIKit.UIView, Microsoft.iOS, Version=17.2.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065]]..ctor(HashSet`1 )
   at System.Collections.Frozen.FrozenSet.CreateFromSet[UIView](HashSet`1 )
   at System.Collections.Frozen.FrozenSet.ToFrozenSet[UIView](IEnumerable`1 , IEqualityComparer`1 )
   at CommunityToolkit.Maui.Core.Views.AlertView.get_Children()
   at CommunityToolkit.Maui.Core.Views.AlertView.Initialize()
   at CommunityToolkit.Maui.Core.Views.Alert.Show()
   at CommunityToolkit.Maui.Alerts.Toast.ShowPlatform(CancellationToken token)
   at CommunityToolkit.Maui.Alerts.Toast.Show(CancellationToken token)
   at KinesiaOne.ViewModels.SettingsPageViewModel.VersionTapped()
```

### Expected Behavior

The toast pop-up should be shown.

### Steps To Reproduce

1. Clone the repository
2. Set the Bundle ID to one that you have a Distribution certificate for
3. Build as Release and push to a device
4. Press the "Click me" button and see the exception and that there is no toast displayed

### Link to public reproduction project repository

https://github.com/vouksh/CommunityToastBug

### Environment

```markdown
- .NET MAUI CommunityToolkit: 7.0.1
- OS: Windows 11 Build 22631.3155 23H2 | Paired to Intel MacBook Pro, macOS 14.3.1 - Xcode 15.2
- .NET MAUI: 8.0.10
- .NET: 8.0.202
- Visual Studio Community 17.9.3
```

### Anything else?

This worked previously before updating Visual Studio and MAUI.

### PR DESCRIPTION
**Pull request by mjo151**
### Description of Change ###

Fixes Snackbar and Toast on iOS release builds

 ### Linked Issues ###

 - Fixes #1752

 ### PR Checklist ###

 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls 

 ### Additional information ###

Removed the use of FrozenSet in the iOS AlertView. FrozenSet was causing the crash on iOS release builds, and it also adds unnecessary overhead when initializing the AlertView.

### CODE REVIEW

**Comment by mjo151** [comment]
@dotnet-policy-service agree

**Comment by brminnick** [comment]
Thanks @mjo151. 

However, this doesn't fix the the two root causes:
1. `<MtouchInterpreter>-all</MtouchInterpreter>` fails to include the `System.Collections.Frozen` namespace: https://github.com/CommunityToolkit/Maui/issues/1752#issuecomment-1999990148
2. The Mono Interpreter is currently required for Release builds: https://github.com/dotnet/maui/issues/13019#issuecomment-1995427494

I think the best path forward for us on the Toolkit is the following:
1. Update our documentation to include the following code as a requirement to use Snackbar + Toast: https://github.com/CommunityToolkit/Maui/issues/1752#issuecomment-2000032581
    ```cs
    <PropertyGroup Condition="$(Configuration.Contains('Release')) And $(TargetFramework.Contains('ios'))">
	    <MtouchInterpreter>-all,System.Collections.Immutable</MtouchInterpreter>
    </PropertyGroup>
    ```
2. Monitor the .NET MAUI / .NET iOS teams progress on the Mono Interpreter: https://github.com/dotnet/maui/issues/13019#issuecomment-1995427494

**Comment by mjo151** [comment]
> Thanks @mjo151.
> 
> However, this doesn't fix the the two root causes:
> 
> 1. `<MtouchInterpreter>-all</MtouchInterpreter>` fails to include the `System.Collections.Immutable` namespace: [[BUG] Toast.Show crashing on iOS Release builds #1752 (comment)](https://github.com/CommunityToolkit/Maui/issues/1752#issuecomment-1999990148)
> 2. The Mono Interpreter is currently required for Release builds: [Consider enabling the interpreter by default for release builds dotnet/maui#13019 (comment)](https://github.com/dotnet/maui/issues/13019#issuecomment-1995427494)
> 
> I think the best path forward for us on the Toolkit team is the following:
> 
> 1. Update our documentation to include the following code as a requirement to use Snackbar + Toast as discussed with @bijington: [[BUG] Toast.Show crashing on iOS Release builds #1752 (comment)](https://github.com/CommunityToolkit/Maui/issues/1752#issuecomment-2000032581)
>    ```cs
>    <PropertyGroup Condition="$(Configuration.Contains('Release')) And $(TargetFramework.Contains('ios'))">
>        <MtouchInterpreter>-all,System.Collections.Immutable</MtouchInterpreter>
>    </PropertyGroup>
>    ```
> 2. Monitor the .NET MAUI / .NET iOS teams progress on the Mono Interpreter: [Consider enabling the interpreter by default for release builds dotnet/maui#13019 (comment)](https://github.com/dotnet/maui/issues/13019#issuecomment-1995427494)

@brminnick 

This PR does fix the root cause of the crash on iOS, which is the use of FrozenSet in the AlertView. The AlertView is used by Snackback and Toast. As you can see from the changes in my PR, the use of FrozenSet is completely unnecessary in this case. I don't think it makes sense to force users to enable the interpreter, which has other implications, when this problem can simply be fixed by removing the use of FrozenSet.

**Comment by brminnick** [comment]
I understand that. This is a workaround to a bug in the .NET Runtime and in the Mono Interpreter.

**Comment by mjo151** [comment]
> I understand that. This is a workaround to a bug in the .NET Runtime and in the Mono Interpreter.

I agree that not being able to use System.Collections.Immutable appears to be a bug in the .NET Runtime and/or the mono AOT compiler. This PR does "workaround" that bug by not using System.Collections.Immutable. However, use of System.Collections.Immutable here was unnecessary anyway and also had negative performance implications on the code.

For various reasons, I am not able to enable the Mono Interpreter for my project. It's very unfortunate if that will be the stance for fixing issues moving forward, especially when there are other options. I really hope you review and consider this PR for integration into the community toolkit. I've been very happy with the toolkit so far, but if using Snackbar and Toast will require enabling the interpreter (e.g. this PR doesn't get approved), I will have to migrate away from using the toolkit.

**Comment by mjo151** [comment]
OK, I understand the issue with breaking the public API. I just pushed another commit that reverts the breaking change to the public API. It still resolves the crash when displaying Snackbar and Toast.

**Review by brminnick — APPROVED** [review]

File: src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ void Initialize()
 			TranslatesAutoresizingMaskIntoConstraints = false
 		};
 
-		foreach (var view in Children)
+		foreach (var view in children)
```
Comment (mattleibow) [inline src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:95]: Will this cause issues if the collection changes during this loop? Before it was a copy, now it is the actual collection and IEnumerable throws if you change it while it is iterating.

Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:95]: It's a possibility, but the likelihood is slim. 

As Vlad pointed out in our standup today that Snackbar + Toast typically only contain 2-3 children, so enumerating through the list should be very quick. The odds of the `List` being modified during enumeration are very low.

Comment (mjo151) [reply src/CommunityToolkit.Maui.Core/Views/Alert/AlertView.macios.cs:95]: Prior to this change, the foreach loop accessed the Children property, which calls children.ToFrozenSet(). ToFrozenSet() has to iterate over the collection.

In either case, AlertView is a UI control and should only ever be modified on the UI thread. If another thread is adding children to the view, that's a bug in the consumer of AlertView.

**Comment by RobertStefancic** [comment]
Is there a chance we get a hotfix release for this soon? Or maybe a preview release?

**Comment by plppp2001** [comment]
> Is there a chance we get a hotfix release for this soon? Or maybe a preview release?

I would like this also, and when could it be released? I can't use pre releases in production env.

**Pull request merged**